src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,241 @@ public static CFProxySettings GetSystemProxySettings ()
 			dict.DangerousRelease ();
 			return new CFProxySettings (dict);
 		}
-		
-#if notyet
-		// FIXME: These require bindings for CFRunLoopSource and CFStreamClientContext
-		
-		public delegate void CFProxyAutoConfigurationResultCallback (IntPtr client, NSArray proxyList, NSError error);
-		
+
+		delegate void CFProxyAutoConfigurationResultCallbackInternal (IntPtr client, IntPtr proxyList, IntPtr error);
+		public delegate void CFProxyAutoConfigurationResultCallback (NSObject client, CFProxy[] proxyList, NSError error);
+		// helper delegate to reuse code
+		delegate IntPtr CreatePACCFRunLoopSource (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context);
+
+		static CFProxy[] ParseProxies (IntPtr proxyList) {
+			CFProxy[] proxies = null;
+			if (proxyList != IntPtr.Zero) {
+				using (var array = new CFArray (proxyList, false)){
+					proxies = new CFProxy [array.Count];
+					for (int i = 0; i < proxies.Length; i++) {
+						var dict = new NSDictionary (array.GetValue (i));
+						proxies[i] = new CFProxy (dict);
+					}
+				}
+			}
+			return proxies;
+		}
+
+		static async Task<(CFProxy[] proxies, NSError error)> ExecutePacCFRunLoopSourceAsync (CreatePACCFRunLoopSource factory, CancellationToken cancellationToken)
+		{
+			CFProxy[] proxies = null;
+			NSError outError = null;
+			await Task.Run (() => {
+				// we need the runloop of THIS thread, so it is important to get it in the correct context
+				var runLoop = CFRunLoop.Current;
+				CFProxyAutoConfigurationResultCallbackInternal cb = delegate (IntPtr client, IntPtr proxyList, IntPtr error) {
+					proxies = ParseProxies (proxyList);
+					if (error != IntPtr.Zero)
+						outError = new NSError (error);
```

Comment:
you parse the proxies even if there's an error - is that intentional ?

Reply:
@spouliot perse, they will be set to null. But I understand it could be misleading, I'll rearrange so that it makes more sense.

Reply:
I have not read the associated documentation - so it's not clear if the proxy (if any) would be usable in case of errors... or if errors also include warnings...

Reply:
According to the documentation if one is not null the other will be (and vice versa).

My vote goes for not adding the IntPtr.Zero checks in this code, since that makes the code smaller. If there's confusion, it can be explained with comments.